File: write_to_lcd/php/WeatherStation.php
<?php

require_once('Tinkerforge/IPConnection.php');
require_once('Tinkerforge/BrickletLCD20x4.php');
require_once('Tinkerforge/BrickletAmbientLight.php');
require_once('Tinkerforge/BrickletAmbientLightV2.php');
require_once('Tinkerforge/BrickletAmbientLightV3.php');
require_once('Tinkerforge/BrickletHumidity.php');
require_once('Tinkerforge/BrickletHumidityV2.php');
require_once('Tinkerforge/BrickletBarometer.php');
require_once('Tinkerforge/BrickletBarometerV2.php');

use Tinkerforge\IPConnection;
use Tinkerforge\BrickletLCD20x4;
use Tinkerforge\BrickletAmbientLight;
use Tinkerforge\BrickletAmbientLightV2;
use Tinkerforge\BrickletAmbientLightV3;
use Tinkerforge\BrickletHumidity;
use Tinkerforge\BrickletHumidityV2;
use Tinkerforge\BrickletBarometer;
use Tinkerforge\BrickletBarometerV2;

class WeatherStation
{
	const HOST = 'localhost';
	const PORT = 4223;

	public function __construct()
    {
		$this->brickletLCD = null;
		$this->brickletAmbientLight = null;
		$this->brickletAmbientLightV2 = null;
		$this->brickletAmbientLightV3 = null;
		$this->brickletHumidity = null;
		$this->brickletHumidityV2 = null;
		$this->brickletBarometer = null;
		$this->brickletBarometerV2 = null;
		$this->ipcon = new IPConnection();

		while(true) {
			try {
				$this->ipcon->connect(self::HOST, self::PORT);
				break;
			} catch(Exception $e) {
				sleep(1);
			}
		}

		$this->ipcon->registerCallback(IPConnection::CALLBACK_ENUMERATE,
		                               array($this, 'cb_enumerate'));
		$this->ipcon->registerCallback(IPConnection::CALLBACK_CONNECTED,
		                               array($this, 'cb_connected'));

		while(true) {
			try {
				$this->ipcon->enumerate();
				break;
			} catch(Exception $e) {
				sleep(1);
			}
		}
	}

	function cb_illuminance($illuminance)
	{
		if($this->brickletLCD != null) {
			$text = sprintf("Illuminanc %6.2f lx", $illuminance/10.0);
			$this->brickletLCD->writeLine(0, 0, $text);
			echo "Write to line 0: $text\n";
		}
	}

	function cb_illuminanceV2($illuminance)
	{
		if($this->brickletLCD != null) {
			$text = sprintf("Illumina %8.2f lx", $illuminance/100.0);
			$this->brickletLCD->writeLine(0, 0, $text);
			echo "Write to line 0: $text\n";
		}
	}

	function cb_illuminanceV3($illuminance)
	{
		if($this->brickletLCD != null) {
			$text = sprintf("Illumina %8.2f lx", $illuminance/100.0);
			$this->brickletLCD->writeLine(0, 0, $text);
			echo "Write to line 0: $text\n";
		}
	}

	function cb_humidity($humidity)
	{
		if($this->brickletLCD != null) {
			$text = sprintf("Humidity   %6.2f %%", $humidity/10.0);
			$this->brickletLCD->writeLine(1, 0, $text);
			echo "Write to line 1: $text\n";
		}
	}

	function cb_humidityV2($humidity)
	{
		if($this->brickletLCD != null) {
			$text = sprintf("Humidity   %6.2f %%", $humidity/100.0);
			$this->brickletLCD->writeLine(1, 0, $text);
			echo "Write to line 1: $text\n";
		}
	}

	function cb_airPressure($airPressure)
	{
		if($this->brickletLCD != null) {
			$text = sprintf("Air Press %7.2f mb", $airPressure/1000.0);
			$this->brickletLCD->writeLine(2, 0, $text);
			echo "Write to line 2: $text\n";

			try {
				$temperature = $this->brickletBarometer->getChipTemperature();
			} catch(Exception $e) {
				echo "Could not get temperature: $e\n";
				return;
			}

			// 0xDF == ° on LCD 20x4 charset
			$text = sprintf("Temperature %5.2f %cC", $temperature/100.0, 0xDF);
			$this->brickletLCD->writeLine(3, 0, $text);
			$text = str_replace(sprintf("%c", 0xDF), '°', $text);
			echo "Write to line 3: $text\n";
		}
	}

	function cb_airPressureV2($airPressure)
	{
		if($this->brickletLCD != null) {
			$text = sprintf("Air Press %7.2f mb", $airPressure/1000.0);
			$this->brickletLCD->writeLine(2, 0, $text);
			echo "Write to line 2: $text\n";

			try {
				$temperature = $this->brickletBarometerV2->getTemperature();
			} catch(Exception $e) {
				echo "Could not get temperature: $e\n";
				return;
			}

			// 0xDF == ° on LCD 20x4 charset
			$text = sprintf("Temperature %5.2f %cC", $temperature/100.0, 0xDF);
			$this->brickletLCD->writeLine(3, 0, $text);
			$text = str_replace(sprintf("%c", 0xDF), '°', $text);
			echo "Write to line 3: $text\n";
		}
	}

	function cb_enumerate($uid, $connectedUid, $position, $hardwareVersion,
	                      $firmwareVersion, $deviceIdentifier, $enumerationType)
	{
		if($enumerationType == IPConnection::ENUMERATION_TYPE_CONNECTED ||
		   $enumerationType == IPConnection::ENUMERATION_TYPE_AVAILABLE) {
			if($deviceIdentifier == BrickletLCD20x4::DEVICE_IDENTIFIER) {
				try {
					$this->brickletLCD = new BrickletLCD20x4($uid, $this->ipcon);
					$this->brickletLCD->clearDisplay();
					$this->brickletLCD->backlightOn();
					echo "LCD 20x4 initialized\n";
				} catch(Exception $e) {
					$this->brickletLCD = null;
					echo "LCD 20x4 init failed: $e\n";
				}
			} else if($deviceIdentifier == BrickletAmbientLight::DEVICE_IDENTIFIER) {
				try {
					$this->brickletAmbientLight = new BrickletAmbientLight($uid, $this->ipcon);
					$this->brickletAmbientLight->setIlluminanceCallbackPeriod(1000);
					$this->brickletAmbientLight->registerCallback(BrickletAmbientLight::CALLBACK_ILLUMINANCE,
					                                              array($this, 'cb_illuminance'));
					echo "Ambient Light initialized\n";
				} catch(Exception $e) {
					$this->brickletAmbientLight = null;
					echo "Ambient Light init failed: $e\n";
				}
			} else if($deviceIdentifier == BrickletAmbientLightV2::DEVICE_IDENTIFIER) {
				try {
					$this->brickletAmbientLightV2 = new BrickletAmbientLightV2($uid, $this->ipcon);
					$this->brickletAmbientLightV2->setConfiguration(BrickletAmbientLightV2::ILLUMINANCE_RANGE_64000LUX,
					                                                BrickletAmbientLightV2::INTEGRATION_TIME_200MS);
					$this->brickletAmbientLightV2->setIlluminanceCallbackPeriod(1000);
					$this->brickletAmbientLightV2->registerCallback(BrickletAmbientLightV2::CALLBACK_ILLUMINANCE,
					                                                array($this, 'cb_illuminanceV2'));
					echo "Ambient Light 2.0 initialized\n";
				} catch(Exception $e) {
					$this->brickletAmbientLight = null;
					echo "Ambient Light 2.0 init failed: $e\n";
				}
			} else if($deviceIdentifier == BrickletAmbientLightV3::DEVICE_IDENTIFIER) {
				try {
					$this->brickletAmbientLightV3 = new BrickletAmbientLightV3($uid, $this->ipcon);
					$this->brickletAmbientLightV3->setConfiguration(BrickletAmbientLightV3::ILLUMINANCE_RANGE_64000LUX,
					                                                BrickletAmbientLightV3::INTEGRATION_TIME_200MS);
					$this->brickletAmbientLightV3->setIlluminanceCallbackConfiguration(1000, false, 'x', 0, 0);
					$this->brickletAmbientLightV3->registerCallback(BrickletAmbientLightV3::CALLBACK_ILLUMINANCE,
					                                                array($this, 'cb_illuminanceV3'));
					echo "Ambient Light 3.0 initialized\n";
				} catch(Exception $e) {
					$this->brickletAmbientLight = null;
					echo "Ambient Light 3.0 init failed: $e\n";
				}
			} else if($deviceIdentifier == BrickletHumidity::DEVICE_IDENTIFIER) {
				try {
					$this->brickletHumidity = new BrickletHumidity($uid, $this->ipcon);
					$this->brickletHumidity->setHumidityCallbackPeriod(1000);
					$this->brickletHumidity->registerCallback(BrickletHumidity::CALLBACK_HUMIDITY,
					                                          array($this, 'cb_humidity'));
					echo "Humidity initialized\n";
				} catch(Exception $e) {
					$this->brickletHumidity = null;
					echo "Humidity init failed: $e\n";
				}
			} else if($deviceIdentifier == BrickletHumidity::DEVICE_IDENTIFIER) {
				try {
					$this->brickletHumidityV2 = new BrickletHumidityV2($uid, $this->ipcon);
					$this->brickletHumidityV2->setHumidityCallbackPeriod(1000, true, 'x', 0, 0);
					$this->brickletHumidityV2->registerCallback(BrickletHumidityV2::CALLBACK_HUMIDITY,
					                                            array($this, 'cb_humidityV2'));
					echo "Humidity 2.0 initialized\n";
				} catch(Exception $e) {
					$this->brickletHumidity = null;
					echo "Humidity 2.0 init failed: $e\n";
				}
			} else if($deviceIdentifier == BrickletBarometer::DEVICE_IDENTIFIER) {
				try {
					$this->brickletBarometer = new BrickletBarometer($uid, $this->ipcon);
					$this->brickletBarometer->setAirPressureCallbackPeriod(1000);
					$this->brickletBarometer->registerCallback(BrickletBarometer::CALLBACK_AIR_PRESSURE,
					                                           array($this, 'cb_airPressure'));
					echo "Barometer initialized\n";
				} catch(Exception $e) {
					$this->brickletBarometer = null;
					echo "Barometer init failed: $e\n";
				}
			} else if($deviceIdentifier == BrickletBarometerV2::DEVICE_IDENTIFIER) {
				try {
					$this->brickletBarometerV2 = new BrickletBarometerV2($uid, $this->ipcon);
					$this->brickletBarometerV2->setAirPressureCallbackConfiguration(1000, false, 'x', 0, 0);
					$this->brickletBarometerV2->registerCallback(BrickletBarometerV2::CALLBACK_AIR_PRESSURE,
					                                             array($this, 'cb_airPressureV2'));
					echo "Barometer 2.0 initialized\n";
				} catch(Exception $e) {
					$this->brickletBarometer = null;
					echo "Barometer 2.0 init failed: $e\n";
				}
			}
		}
	}

	function cb_connected($connectedReason)
	{
		if($connectedReason == IPConnection::CONNECT_REASON_AUTO_RECONNECT) {
			echo "Auto Reconnect\n";

			while(true) {
				try {
					$this->ipcon->enumerate();
					break;
				} catch(Exception $e) {
					sleep(1);
				}
			}
		}
	}
}

$weatherStation = new WeatherStation();
echo "Press ctrl+c to exit\n";
$weatherStation->ipcon->dispatchCallbacks(-1);

?>

File: write_to_lcd/python/weather_station.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import socket
import sys
import time
import math
import logging as log
log.basicConfig(level=log.INFO)

from tinkerforge.ip_connection import IPConnection
from tinkerforge.ip_connection import Error
from tinkerforge.bricklet_lcd_20x4 import BrickletLCD20x4
from tinkerforge.bricklet_ambient_light import BrickletAmbientLight
from tinkerforge.bricklet_ambient_light_v2 import BrickletAmbientLightV2
from tinkerforge.bricklet_ambient_light_v3 import BrickletAmbientLightV3
from tinkerforge.bricklet_humidity import BrickletHumidity
from tinkerforge.bricklet_humidity_v2 import BrickletHumidityV2
from tinkerforge.bricklet_barometer import BrickletBarometer
from tinkerforge.bricklet_barometer_v2 import BrickletBarometerV2

class WeatherStation:
    HOST = "localhost"
    PORT = 4223

    ipcon = None
    lcd = None
    al = None
    al_v2 = None
    al_v3 = None
    hum = None
    hum_v2 = None
    baro = None
    baro_v2 = None

    def __init__(self):
        self.ipcon = IPConnection()
        while True:
            try:
                self.ipcon.connect(WeatherStation.HOST, WeatherStation.PORT)
                break
            except Error as e:
                log.error('Connection Error: ' + str(e.description))
                time.sleep(1)
            except socket.error as e:
                log.error('Socket error: ' + str(e))
                time.sleep(1)

        self.ipcon.register_callback(IPConnection.CALLBACK_ENUMERATE,
                                     self.cb_enumerate)
        self.ipcon.register_callback(IPConnection.CALLBACK_CONNECTED,
                                     self.cb_connected)

        while True:
            try:
                self.ipcon.enumerate()
                break
            except Error as e:
                log.error('Enumerate Error: ' + str(e.description))
                time.sleep(1)

    def cb_illuminance(self, illuminance):
        if self.lcd is not None:
            text = 'Illuminanc %6.2f lx' % (illuminance/10.0)
            self.lcd.write_line(0, 0, text)
            log.info('Write to line 0: ' + text)

    def cb_illuminance_v2(self, illuminance):
        if self.lcd is not None:
            text = 'Illumina %8.2f lx' % (illuminance/100.0)
            self.lcd.write_line(0, 0, text)
            log.info('Write to line 0: ' + text)

    def cb_illuminance_v3(self, illuminance):
        if self.lcd is not None:
            text = 'Illumina %8.2f lx' % (illuminance/100.0)
            self.lcd.write_line(0, 0, text)
            log.info('Write to line 0: ' + text)

    def cb_humidity(self, humidity):
        if self.lcd is not None:
            text = 'Humidity   %6.2f %%' % (humidity/10.0)
            self.lcd.write_line(1, 0, text)
            log.info('Write to line 1: ' + text)

    def cb_humidity_v2(self, humidity):
        if self.lcd is not None:
            text = 'Humidity   %6.2f %%' % (humidity/100.0)
            self.lcd.write_line(1, 0, text)
            log.info('Write to line 1: ' + text)

    def cb_air_pressure(self, air_pressure):
        if self.lcd is not None:
            text = 'Air Press %7.2f mb' % (air_pressure/1000.0)
            self.lcd.write_line(2, 0, text)
            log.info('Write to line 2: ' + text)

            try:
                temperature = self.baro.get_chip_temperature()
            except Error as e:
                log.error('Could not get temperature: ' + str(e.description))
                return

            # \xDF == ° on LCD 20x4 charset
            text = 'Temperature %5.2f \xDFC' % (temperature/100.0)
            self.lcd.write_line(3, 0, text)
            log.info('Write to line 3: ' + text.replace('\xDF', '°'))

    def cb_air_pressure_v2(self, air_pressure):
        if self.lcd is not None:
            text = 'Air Press %7.2f mb' % (air_pressure/1000.0)
            self.lcd.write_line(2, 0, text)
            log.info('Write to line 2: ' + text)

            try:
                temperature = self.baro_v2.get_temperature()
            except Error as e:
                log.error('Could not get temperature: ' + str(e.description))
                return

            # \xDF == ° on LCD 20x4 charset
            text = 'Temperature %5.2f \xDFC' % (temperature/100.0)
            self.lcd.write_line(3, 0, text)
            log.info('Write to line 3: ' + text.replace('\xDF', '°'))

    def cb_enumerate(self, uid, connected_uid, position, hardware_version,
                     firmware_version, device_identifier, enumeration_type):
        if enumeration_type == IPConnection.ENUMERATION_TYPE_CONNECTED or \
           enumeration_type == IPConnection.ENUMERATION_TYPE_AVAILABLE:
            if device_identifier == BrickletLCD20x4.DEVICE_IDENTIFIER:
                try:
                    self.lcd = BrickletLCD20x4(uid, self.ipcon)
                    self.lcd.clear_display()
                    self.lcd.backlight_on()
                    log.info('LCD 20x4 initialized')
                except Error as e:
                    log.error('LCD 20x4 init failed: ' + str(e.description))
                    self.lcd = None
            elif device_identifier == BrickletAmbientLight.DEVICE_IDENTIFIER:
                try:
                    self.al = BrickletAmbientLight(uid, self.ipcon)
                    self.al.set_illuminance_callback_period(1000)
                    self.al.register_callback(self.al.CALLBACK_ILLUMINANCE,
                                              self.cb_illuminance)
                    log.info('Ambient Light initialized')
                except Error as e:
                    log.error('Ambient Light init failed: ' + str(e.description))
                    self.al = None
            elif device_identifier == BrickletAmbientLightV2.DEVICE_IDENTIFIER:
                try:
                    self.al_v2 = BrickletAmbientLightV2(uid, self.ipcon)
                    self.al_v2.set_configuration(self.al_v2.ILLUMINANCE_RANGE_64000LUX,
                                                 self.al_v2.INTEGRATION_TIME_200MS)
                    self.al_v2.set_illuminance_callback_period(1000)
                    self.al_v2.register_callback(self.al_v2.CALLBACK_ILLUMINANCE,
                                                 self.cb_illuminance_v2)
                    log.info('Ambient Light 2.0 initialized')
                except Error as e:
                    log.error('Ambient Light 2.0 init failed: ' + str(e.description))
                    self.al_v2 = None
            elif device_identifier == BrickletAmbientLightV3.DEVICE_IDENTIFIER:
                try:
                    self.al_v3 = BrickletAmbientLightV3(uid, self.ipcon)
                    self.al_v3.set_configuration(self.al_v3.ILLUMINANCE_RANGE_64000LUX,
                                                 self.al_v3.INTEGRATION_TIME_200MS)
                    self.al_v3.set_illuminance_callback_configuration(1000, False, 'x', 0, 0)
                    self.al_v3.register_callback(self.al_v3.CALLBACK_ILLUMINANCE,
                                                 self.cb_illuminance_v3)
                    log.info('Ambient Light 3.0 initialized')
                except Error as e:
                    log.error('Ambient Light 3.0 init failed: ' + str(e.description))
                    self.al_v3 = None
            elif device_identifier == BrickletHumidity.DEVICE_IDENTIFIER:
                try:
                    self.hum = BrickletHumidity(uid, self.ipcon)
                    self.hum.set_humidity_callback_period(1000)
                    self.hum.register_callback(self.hum.CALLBACK_HUMIDITY,
                                               self.cb_humidity)
                    log.info('Humidity initialized')
                except Error as e:
                    log.error('Humidity init failed: ' + str(e.description))
                    self.hum = None
            elif device_identifier == BrickletHumidityV2.DEVICE_IDENTIFIER:
                try:
                    self.hum_v2 = BrickletHumidityV2(uid, self.ipcon)
                    self.hum_v2.set_humidity_callback_configuration(1000, True, 'x', 0, 0)
                    self.hum_v2.register_callback(self.hum_v2.CALLBACK_HUMIDITY,
                                                  self.cb_humidity_v2)
                    log.info('Humidity 2.0 initialized')
                except Error as e:
                    log.error('Humidity 2.0 init failed: ' + str(e.description))
                    self.hum_v2 = None
            elif device_identifier == BrickletBarometer.DEVICE_IDENTIFIER:
                try:
                    self.baro = BrickletBarometer(uid, self.ipcon)
                    self.baro.set_air_pressure_callback_period(1000)
                    self.baro.register_callback(self.baro.CALLBACK_AIR_PRESSURE,
                                                self.cb_air_pressure)
                    log.info('Barometer initialized')
                except Error as e:
                    log.error('Barometer init failed: ' + str(e.description))
                    self.baro = None
            elif device_identifier == BrickletBarometerV2.DEVICE_IDENTIFIER:
                try:
                    self.baro_v2 = BrickletBarometerV2(uid, self.ipcon)
                    self.baro_v2.set_air_pressure_callback_configuration(1000, False, 'x', 0, 0)
                    self.baro_v2.register_callback(self.baro_v2.CALLBACK_AIR_PRESSURE,
                                                   self.cb_air_pressure_v2)
                    log.info('Barometer initialized')
                except Error as e:
                    log.error('Barometer init failed: ' + str(e.description))
                    self.baro_v2 = None

    def cb_connected(self, connected_reason):
        if connected_reason == IPConnection.CONNECT_REASON_AUTO_RECONNECT:
            log.info('Auto Reconnect')

            while True:
                try:
                    self.ipcon.enumerate()
                    break
                except Error as e:
                    log.error('Enumerate Error: ' + str(e.description))
                    time.sleep(1)

if __name__ == "__main__":
    log.info('Weather Station: Start')

    weather_station = WeatherStation()

    if sys.version_info < (3, 0):
        input = raw_input # Compatibility for Python 2.x
    input('Press key to exit\n')

    if weather_station.ipcon != None:
        weather_station.ipcon.disconnect()

    log.info('Weather Station: End')
